python/lbann/modules/transformer.py

```diff
@@ -91,26 +108,118 @@ def forward(self, queries, keys, values, mask=None):
         self.instance += 1
         name = f'{self.name}_instance{self.instance}'
 
-        # Apply fully-connected layers to input sequences
-        queries_fc = lbann.ChannelwiseFullyConnected(
-            queries,
-            weights=self.query_weights,
+        if self.self_attention:
+            # If self-attention, multiply with stacked matrix
+            assert queries is keys and keys is values
+
+            qkv_fc = lbann.ChannelwiseFullyConnected(
+                queries,
+                weights=self.qkv_weights,
+                output_channel_dims=[self.embed_dim * 3],
+                name=f'{name}_qkv_fc',
+                bias=True,
+                transpose=False
+            )
+
+            # Unstack
+            qkv_slice = lbann.Slice(qkv_fc,
+                                    axis=1,
+                                    slice_points=[
+                                        0, self.embed_dim, 2 * self.embed_dim,
+                                        3 * self.embed_dim
+                                    ])
+            queries_fc = lbann.Identity(qkv_slice)
+            keys_fc = lbann.Identity(qkv_slice)
+            values_fc = lbann.Identity(qkv_slice)
```

Comment:
Not a criticism, but we should make a note that this notation for slicing is quite unintuitive. Maybe we should revisit it at some point.